ECS chapter for new book (#182)
bevyengine/bevy-website: content/learn/book/ecs/_index.md

* we access data from this database using **queries**, which fetch entities with the specified components
* the primary key of this database is the [`Entity`](https://docs.rs/bevy/latest/bevy/ecs/entity/struct.Entity.html) identifier, which can use to look up specific entities using [`Query::get(my_entity)`](https://docs.rs/bevy/latest/bevy/ecs/prelude/struct.Query.html#method.get)

Of course, this database is not very well-normalized: not all entities will have every component!
We can use this to specialize behavior between entities: only performing work on them in our systems if they have the correct combination of components.
You don't want to apply gravity to entities without a position in your world, and you're only interested in using the UI layout algorithm to control the layout of UI entities!

When we want to go beyond this tabular data storage, we can use **resources**: global singletons which store data in monolithic blobs.
You might use resources to store one-off bits of state like the game's score, use it to interface with other libraries, or store secondary data structures like indexes to augment your use of entity-component data.

> Just like with components on entities, resources are accessed by type and you can only have one resource of each Rust type.

Author: I think this whole line is a poor fit here. The point is to introduce the concepts, not flood the reader with details.

In order to actually perform logic on all of this data, we must use systems, which are automatically run by our scheduler.

> **Systems** are Rust functions which request specific data from the world, as declared in their system parameters (function arguments): generally resources and entities that have a particular combination of components using queries.

Reviewer: - I agree that "using queries" is weird here. However, I think that whole elaboration of *what* data systems might request could be left out, since..

- I think this explanation doesn't really give the reader an idea of what the *purpose* of a system is. 

  As a reader, why do I care about Systems? What does this feature let me do?
  
  My suggestion would be to mention the purpose a bit more: perhaps something like:
  
  > **Systems** are Rust functions that request data from the World, in order to perform tasks - often a system is a procedure that specifies how entities may interact. They prescribe all of the rules and behaviours of the World.

> **Systems** are Rust functions which request specific data from the world, as declared in their system parameters (function arguments): generally resources and entities that have a particular combination of components using queries.

Author: Adapted your suggestion <3 I agree that there needed to be more *intent* present.

> Systems are Rust functions that request data from the `World` (such as resources or entities with particular components) in order to perform tasks.
> All of the rules and behaviours of our game are governed by systems.

Once the systems are added to our app the **scheduler** takes in this information and automatically runs our systems: typically once during each pass of the **game loop**.

Bevy's scheduler is remarkably powerful: it uses the information about data access defined in our system parameters to automatically run systems in parallel.
By default, every system in a stage runs in parallel with every other system in that stage (as long as threads exist to take the work): the only rule is that systems which have the ability to *write* to a particular piece of data (such as a resource) cannot be run at the same time as other systems which read or write to that same data.
In Rust, this corresponds to mutable access, as declared by the use of `Query<&mut MyComponent>` or `ResMut<MyResource>`.

On the next page, we'll create a simple "game" using the ECS so you can see how this all fits together.

## Rust tips and tricks for Bevy's ECS

> If you're new to Rust, there are a few standard tricks and idioms that are used by Bevy's ECS that are worth learning about.

Reviewer 2: > there are a few standard tricks and idioms that are used by Bevy's ECS that are worth learning about.

I found a couple of potential problems in this sentence:

1. The word _tricks_ is a bit too colloquial and lacks strong meaning. Using _patterns_ instead seems more appropriate.
2. The phrase "that are used by Bevy's ECS" while maybe true, it's probably not what you actually meant. Did you mean something like "that you can adopt to use Bevy's ECS at it's fullest" instead?

Author: > If you're new to Rust, there are a few standard patterns and pitfalls that are particularly relevant to Bevy's ECS.

Feel free to come back to this section later as you learn more about Bevy!

### Tuple structs

> When working with tuple structs like `struct Score(u64)` or `struct Position(u8, u8)`, `.0`, `.1` and so on are used to access fields, counting up from 0 in the order they were defined.

Reviewer 2: It's not clear where the definition of the tuple struct ends and where the field access syntax starts. Separating the two concepts in two separate sentences may be the easiest solution.

> `let foo.0 = bar` can seem opaque when you first read it, but in Rust it just means that you're assigning `bar` to the first field of the tuple struct `foo`.

Reviewer 2: I would avoid _foo_, _bar_ and similar meaningless identifiers. Names that reflect the role of a thing are much more better IMO. Example:

```
my_tuple.0 = new_value
```

Note that I removed `let` as it should be incorrect.

1. The value returned in a system when you ask for `Res<Score>` is not a plain `Score`, it is a `Res<Score>`. This is a resource that stores the score, similar to an ordinary `&` pointer, but with some special behavior (like change detection).
2. Because `Res` and other smart pointers implement the `Deref` trait, Rust will attempt to automatically dereference it to match required types when it can.
   1. This is commonly done when accessing fields and methods: calling `.title` on `Res<WindowDescriptor>` will grab the title field of `WindowDescriptor`, not the wrapping `Res<WindowDescriptor>` as the `Res` type has no field named `title`.
3. When your types cannot be automatically dereferenced, you will need to manually dereference out of the outer smart pointer using `*`.
   1. This is commonly needed when assigning a value to your data (rather than just one of its fields), or when using `match`.

### Initializing structs with some default values

Bevy often requires defining complex structs when creating entities with bundles or configuring the app with resources.
In many cases, we want to use *some* default fields, but not all of them.

> As Rust does not have default arguments, we instead use [struct update syntax](https://doc.rust-lang.org/book/ch05-01-defining-structs.html#creating-instances-from-other-instances-with-struct-update-syntax) in combination with the [`default()`](https://doc.rust-lang.org/std/default/trait.Default.html) method from the `Default` trait to allow us to quickly set some values while leaving others unchanged.

Reviewer 2: ```suggestion
As Rust does not have default arguments, we instead use [struct update syntax (SUS)](https://doc.rust-lang.org/book/ch05-01-defining-structs.html#creating-instances-from-other-instances-with-struct-update-syntax) in combination with the [`default()`](https://doc.rust-lang.org/std/default/trait.Default.html) method from the `Default` trait to allow us to quickly set some values while leaving others unchanged.
```

It was stronger than me lol. Please don't accept it 😝

Reviewer: Since `default` is the *only* required method of the `Default` trait, this seems needlessly wordy. My suggestion would be something along the lines of:

> ".. we instead use struct update syntax in combination with [`Default::default()`](https://doc.rust-lang.org/std/default/trait.Default.html), which allows us to succinctly specify only the values we are interested in."

Author: > ".. we instead use struct update syntax in combination with Default::default(), which allows us to succinctly specify only the values we are interested in."

Stealing this bit :)

> As Rust does not have default arguments, we instead use [struct update syntax](https://doc.rust-lang.org/book/ch05-01-defining-structs.html#creating-instances-from-other-instances-with-struct-update-syntax) in combination with the [`default()`](https://doc.rust-lang.org/std/default/trait.Default.html) method from the `Default` trait to allow us to quickly set some values while leaving others unchanged.

Author: I actually don't think this is grammatically correct. Explaining precisely *why* as a native English speaker is hard though 😅

```rust
// We want to quickly generate new enemies from this component bundle
// By using some default fields
#[derive(Bundle, Default)]
struct EnemyBundle {
	enemy: Enemy,
	life: Life,
	attack: Attack,
	defense: Defense,
}

// Dataless unit structs automatically have a default value
// Since only one possible value exists
struct Enemy;

struct Life(u32);

// Manually implementing Default for one of our component types
// allowing us to control the starting value
impl Default for Life {
	fn default() -> Self {
		Life(100)
	}
}

// These components will have a default value of 0, as u32::default() == 0
#[derive(Default)]
struct Defense(u32);

#[derive(Default)]
struct Attack(u32);

let my_enemy = EnemyBundle {
	attack: 10,
	defense: 5,
	// All other fields are initialized from their default value
	// setting enemy: Enemy and life: Life(100)
	..Default::default()
}

> ```

Reviewer 2: The example should just focus on the Struct Update Syntax, not bleed into Bevy features. Here the concept of _bundles_ is added, unnecessarily increasing complexity (especially since bundles are not introduced yet). A pure Rust example would be much more well understood.

Basically, just show the definition of a struct with many fields (not too many or the code explodes in length; 4 fields are enough). Say that the struct implements `Default` (don't show the full impl though, it's too much code and it's irrelevant), and then create an instance of the struct by filling a couple of fields and then use `..Default::default()`

Author: Agreed: this has been simplified and I left a note below to talk about its relevance to bundles.

### Type aliases don't play nice

> Bey's ECS uses Rust's type system to dispatch data to our systems as requested.

Reviewer 2: ```suggestion
Bevy's ECS uses Rust's type system to dispatch data to our systems as requested.
```

Typo correction.

"One of each Rust type" has somewhat surprising consequences for new Rust users though: [type aliases](https://doc.rust-lang.org/reference/items/type-aliases.html) will not result in unique types, and should generally not be used to define resource or component types.

```rust
use std::any::type_name;

// Our first component type
struct Life(u32);

// An attempt at a second component type, using a type alias
type Health = Life;

// Unfortunately, these two types share the same name :(
// Attempting to insert both Life and Health would result in overwritten values,
// and any query for one would return either
assert_eq!(type_name(Life(42)), type_name(Health(42));
```

Instead, you have to define entirely new types for each component or resource you wish to use,
commonly using the [`newtype` pattern](https://doc.rust-lang.org/book/ch19-03-advanced-traits.html#using-the-newtype-pattern-to-implement-external-traits-on-external-types) to wrap a single value of some common or external type.

```rust
use std::any::type_name;

struct Life(u32);
struct Health(u32);

// Our components have different types now!
assert!(type_name(Life(42)) != type_name(Health(42));

> ```

Reviewer 2: The use of type ids instead of type names feels more appropriate to me, since that is what really distinguished two different types.

Author: Absolutely. Changing.

* the primary key of this database is the [`Entity`](https://docs.rs/bevy/latest/bevy/ecs/entity/struct.Entity.html) identifier, which can be used to look up specific entities using [`Query::get(my_entity)`](https://docs.rs/bevy/latest/bevy/ecs/prelude/struct.Query.html#method.get)

Of course, this database is not very well-normalized: not all entities will have every component!
We can use this fact to specialize behavior between entities: systems only perform work on entities with the correct combination of components.
You don't want to apply gravity to entities without a position in your world, and you're only interested in using the UI layout algorithm to control the layout of UI entities!

When we want to go beyond this tabular data storage, we can use **resources**: global singletons which store data in monolithic blobs.
You might use resources to store one-off bits of state like the game's score, use it to interface with other libraries, or store secondary data structures like indexes to augment your use of entity-component data.

In order to actually manipulate all of this data in interesting ways, we must use systems.

> **Systems** are Rust functions which request specific data from the world, as declared in their system parameters (function parameters): generally resources and entities that have a particular combination of components using queries.

Reviewer 3: I tried to make it clearer but I may have lost some information.

In Rust, this corresponds to mutable access, as declared by the use of `Query<&mut MyComponent>` or `ResMut<MyResource>`.

On the next page, we'll create a simple "game" using the ECS so you can see how this all fits together.

## Rust tips and tricks for Bevy's ECS

If you're new to Rust, there are a few standard patterns and pitfalls that are particularly relevant to Bevy's ECS.
Feel free to come back to this section later as you learn more about Bevy!

### Tuple structs

Tuple structs, defined like `struct Score(u64);` or `struct Position(u8, u8);`, are structs with unnamed fields, accessed by the order in which they were defined.
`.0`, `.1` and so on are used to access fields, counting up from 0 in the order they were declared.

`my_position.0 = 42` can seem opaque when you first read it, but in Rust it just means that you're assigning `42` to the first field of the tuple struct `my_position` (by convention, the x coordinate).
Using ordinary structs with named fields rather than tuple structs (even for simple components and resources!) can go a long way to improving the clarity of your code by providing more descriptive field names.

> In this example, `my_position.x = 42` immediately communicates intent, rather than relying on the end user to remember the convention used.

Reviewer 3: ```suggestion
In this example, `my_position.x = 42` immediately communicates intent, rather than relying on future you to remember the convention used.
```
Minor nitpick: the end-user of a game usually doesn't look at the code.

Reviewer 4: ```suggestion
In this example, `my_position.x = 42` immediately communicates intent, rather than relying on future developers to remember the convention used.
```
I think it should be 'future developers' because more than 1 person may be working on the code.

### Smart pointers

There are a few points to be aware of when working with "smart pointers" like components and resources:

> 1. The value returned in a system when you ask for `Res<Score>` is not a plain `Score`, it is a `Res<Score>`. This is a resource that stores the score, similar to an ordinary `&` pointer, but with some special behavior (like change detection).

Reviewer 3: ```suggestion
1. The value passed to a system when you ask for `Res<Score>` is not a plain `Score`: it is a `Res<Score>`. This is a resource that stores the score, similar to an ordinary `&` pointer, but with some special behavior (like change detection).
```

2. Because `Res` and other smart pointers implement the `Deref` trait, Rust will attempt to automatically dereference it to match required types when it can.
   1. This is commonly done when accessing fields and methods: calling `.title` on `Res<WindowDescriptor>` will grab the title field of `WindowDescriptor`, not the wrapping `Res<WindowDescriptor>` as the `Res` type has no field named `title`.

> 3. When your types cannot be automatically dereferenced, you will need to manually dereference out of the outer smart pointer using `*`.

Reviewer 3: ```suggestion
3. When your types cannot be automatically dereferenced, you will need to manually dereference the outer smart pointer (i.e. `Res`) using `*`.
```
A small clarification to make it very clear.

>    1. This is commonly needed when assigning a value to your data (rather than just one of its fields), or when using `match`.

Reviewer 3: ```suggestion
   1. This is commonly needed when assigning a value to your data (rather than just one of its fields), or using `match`.
```

### Initializing structs with some default values

Bevy often requires defining complex structs when creating entities with bundles or configuring the app with resources.
In many cases, we want to use *some* default fields, but not all of them.
As Rust does not have default arguments, we instead use [struct update syntax](https://doc.rust-lang.org/book/ch05-01-defining-structs.html#creating-instances-from-other-instances-with-struct-update-syntax) in combination with the [`Default::default()`](https://doc.rust-lang.org/std/default/trait.Default.html) method to allow us to quickly set some values while leaving others unchanged.

```rust
struct CombatStats {
	attack: u8,
	defense: u8,
	speed: u8,
}

impl Default for CombatStats {
	// This method is being used to create an initial struct
	// whose values we copy and then override
	fn default() -> Self {
		CombatStats {
			attack: 100,
			defense: 100,
			speed: 100,
		}
	}
}

let my_stats = CombatStats {
	attack: 150,
	defense: 50,
	..Default::default()
}
```

This pattern is commonly used when working with bundles: groups of components typically used for entity initialization.

### Type aliases don't play nice

> Bey's ECS uses Rust's type system to dispatch data to our systems as requested.

Reviewer 2: ```suggestion
Bevy's ECS uses Rust's type system to dispatch data to our systems as requested.
```

You forgot to commit this typo fix.

> Bevy's ECS uses Rust's type system to dispatch data to our systems as requested.

Reviewer 3: ```suggestion
Bevy's ECS uses Rust's type system to dispatch data to our systems as required.
```

> This is very convenient for safety and ergonomic reasons, but means that only one resource of each Rust type can exist at once, and only one component of each type can be stored on each entity.

Reviewer 3: ```suggestion
This is very convenient for safety and ergonomic reasons but means that no two resources can have the same Rust type, and entities can, at most, only have one instance of each type of component.
```
In my opinion, this is clearer, but I might be wrong.

> "One of each Rust type" has somewhat surprising consequences for new Rust users though: [type aliases](https://doc.rust-lang.org/reference/items/type-aliases.html) will not result in unique types, and should generally not be used to define resource or component types.

Reviewer 3: ```suggestion
No two resources having the same Rust type has somewhat surprising consequences for new Rust users: [type aliases](https://doc.rust-lang.org/reference/items/type-aliases.html) will not result in unique types, and should generally not be used to define resource or component types.
```

```rust
use std::any::type_name;

// Our first component type
struct Life(u32);

// An attempt at a second component type, using a type alias
type Health = Life;

> // Unfortunately, these two types share the same name :(

Reviewer 2: ```suggestion
// Unfortunately, these two types share the same id :(
```

* the primary key of this database is the {{rust_type(type="struct" crate="bevy_ecs" name="Entity")}} identifier, which can be used to look up specific entities using {{rust_type(type="struct" crate="bevy_ecs" name="Query" method = "get")}}

Of course, this database is [very ragged](https://www.transdatasolutions.com/what-is-ragged-data/): not all entities will have every component!
We can use this fact to specialize behavior between entities: systems only perform work on entities with the correct combination of components.
You don't want to apply gravity to entities without a position in your world, and you're only interested in using the UI layout algorithm to control the layout of UI entities!

When we want to go beyond this tabular data storage, we can use **resources**: global singletons which store data, each in their own monolithic blob.
You might use resources to interface with other libraries, store unique bits of state like the game's score, or store secondary data structures like indexes to augment your use of entity-component data.

In order to manipulate and act on this data, we must use systems.

> **Systems** are Rust functions that request specific data, such as resources and entities, from the {{rust_type(type="struct" crate="bevy_ecs" name="World")}}. They define a query in their parameters (arguments) that selects data with a particular combination of components.

Author: I like the bullet points a lot. I'm going to steal from this.

When we need to access data in complex, cross-cutting ways that are not cleanly modelled by our systems' function signatures, we can defer the work until we have exclusive access to the entire world's data: executing **commands** generated in earlier systems at the end of each stage or performing complex logic (like saving the entire game) in our own **exclusive systems**.
You will first encounter this when spawning and despawning entities: we have no way of knowing precisely which other components our entities might have, and so we are forced to wait until we can ensure that we can safely write to *all* component data at once.

## ECS by example

Before we dive into the details of each of these features, let's take a quick look at a simple but fun whack-a-mole game that you can run and play.
Unsurprisingly, the different parts of the ECS tend to be closely linked: components are not very useful without a way to spawn entities and systems that run our logic are very dull if we can't discuss the data they can access.
The details of each part are more easily grasped if you have a basic sense of the whole.

```rust

> use bevy::prelude::*;

Author: I agree, this example is way too complex for a new user.
